execution/src/transaction.rs

```diff
@@ -25,45 +29,131 @@ pub const L2_LOWEST_ADDRESS: Address = address!("4200000000000000000000000000000
+    fn encode(&self, out: &mut dyn alloy::rlp::BufMut) {
+        let envelope: TxEnvelope = self.clone().into();
+        envelope.encode(out);
+    }
+}
+
 pub trait WrapReceipt {
     fn wrap_receipt(&self, receipt: Receipt, bloom: Bloom) -> OpReceiptEnvelope;
 }
 
-impl WrapReceipt for OpTxEnvelope {
+impl WrapReceipt for NormalizedExtendedTxEnvelope {
     fn wrap_receipt(&self, receipt: Receipt, bloom: Bloom) -> OpReceiptEnvelope {
         match self {
-            Self::Legacy(_) => OpReceiptEnvelope::Legacy(ReceiptWithBloom {
-                receipt,
-                logs_bloom: bloom,
-            }),
-            Self::Eip1559(_) => OpReceiptEnvelope::Eip1559(ReceiptWithBloom {
-                receipt,
-                logs_bloom: bloom,
-            }),
-            Self::Eip2930(_) => OpReceiptEnvelope::Eip2930(ReceiptWithBloom {
-                receipt,
+            Self::Canonical(norm_tx) => match norm_tx.tx_type {
+                UmiTxType::Legacy => OpReceiptEnvelope::Legacy(ReceiptWithBloom {
+                    receipt,
+                    logs_bloom: bloom,
+                }),
+                UmiTxType::Eip2930 => OpReceiptEnvelope::Eip2930(ReceiptWithBloom {
+                    receipt,
+                    logs_bloom: bloom,
+                }),
+                UmiTxType::Eip1559 => OpReceiptEnvelope::Eip1559(ReceiptWithBloom {
+                    receipt,
+                    logs_bloom: bloom,
+                }),
+            },
+            Self::DepositedTx(dep_tx) => OpReceiptEnvelope::Deposit(OpDepositReceiptWithBloom {
+                receipt: OpDepositReceipt {
+                    inner: receipt,
+                    // Per OP stack spec <https://specs.optimism.io/protocol/deposits.html#execution>
+                    deposit_nonce: Some(dep_tx.nonce()),
+                    deposit_receipt_version: Some(DEPOSIT_RECEIPT_VERSION),
+                },
                 logs_bloom: bloom,
             }),
-            Self::Deposit(dep) => {
-                OpReceiptEnvelope::Deposit(OpDepositReceiptWithBloom {
-                    receipt: OpDepositReceipt {
-                        inner: receipt,
-                        // Per OP stack spec <https://specs.optimism.io/protocol/deposits.html#execution>
-                        deposit_nonce: Some(dep.nonce()),
-                        deposit_receipt_version: Some(DEPOSIT_RECEIPT_VERSION),
-                    },
-                    logs_bloom: bloom,
-                })
-            }
-            _ => unreachable!("Wrapping of other tx types not supported"),
         }
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Deserialize, Serialize)]
 pub enum NormalizedExtendedTxEnvelope {
     Canonical(NormalizedEthTransaction),
-    DepositedTx(TxDeposit),
+    DepositedTx(Sealed<TxDeposit>),
+}
+
+// TODO: avoid conversion to op-alloy and do it internally
+impl Decodable for NormalizedExtendedTxEnvelope {
```

Comment:
- [ ] Similarly with the serde traits, I wonder if this impl can be avoided by keeping op-alloy types at all IO boundaries.